### ISSUE TITLE
Signed URL generation is really slow

### ISSUE THREAD

**Issue by AndrewVos** [GoogleCloudPlatform/google-cloud-ruby#1359]
I'm migrating over from S3 and I was really surprised to see that signed url generation is extremely slow with this library. This must be a bug? I'm getting one of these for every signed url:

```
Sending HTTP get https://www.googleapis.com/storage/v1/b/REDACTED?
200
#<Hurley::Response GET https://www.googleapis.com/storage/v1/b/REDACTED == 200 (586 bytes) 217ms>
Success - #<Google::Apis::StorageV1::Bucket:REDACTED
 @cors_configurations=
  [#<Google::Apis::StorageV1::Bucket::CorsConfiguration:REDACTED
    @http_method=["PUT"],
    @max_age_seconds=3600,
    @origin=["*"],
    @response_header=["Content-Type", "Access-Control-Allow-Origin"]>],
 @etag="CAM=",
 @id="REDACTED",
 @kind="storage#bucket",
 @location="US",
 @metageneration="3",
 @name="REDACTED",
 @project_number="REDACTED",
 @self_link="https://www.googleapis.com/storage/v1/b/REDACTED",
 @storage_class="MULTI_REGIONAL",
 @time_created=Sun, 19 Mar 2017 15:55:07 +0000,
 @updated=Sun, 19 Mar 2017 16:22:25 +0000>

```

Is this expected?

This is my code:

```ruby

  def bucket
    @storage ||= Google::Cloud::Storage.new
    @bucket ||= @storage.bucket(ENV.fetch('GOOGLE_CLOUD_STORAGE_BUCKET'))
  end

  def thumbnail_url
    bucket.signed_url(
      thumbnail_name,
      method: 'GET',
      expires: 300
    )
  end

```

**Comment by quartzmo** [comment]
Hi @AndrewVos, thanks for reporting this issue.

Have you tried [Creating Signed URLs with gsutil](https://cloud.google.com/storage/docs/access-control/create-signed-urls-gsutil)? I am wondering how the response time would compare.

**Comment by AndrewVos** [comment]
@quartzmo yeah I have and if I recall correctly there was some sort of request happening with that too. Strange.

Anyway, this is way faster: https://blog.oestrich.org/2016/01/signing-google-cloud-storage-urls/

**Comment by AndrewVos** [comment]
Oh also another issue, if `content_type` isn't specified in `PUT` then the signature will not match.

Not sure if that is a bug or just lack of documentation.

**Comment by blowmage** [comment]
> Is this expected?

It depends on how the code you've shared is being used. If the `bucket` instance object is being created new each time, then yes it will make an API request for the Bucket data.

> Oh also another issue, if content_type isn't specified in PUT then the signature will not match.
> 
> Not sure if that is a bug or just lack of documentation.

The [GCS docs](https://cloud.google.com/storage/docs/access-control/create-signed-urls-program) seem to state that pretty clearly.

**Comment by blowmage** [comment]
@quartzmo Should we add `Project#signed_url` that accepts both `bucket` and `path`, similar to [`Bucket#signed_url`](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-storage/v0.24.0/google/cloud/storage/bucket?method=signed_url-instance) and [`File#signed_url`](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-storage/v0.24.0/google/cloud/storage/file?method=signed_url-instance)?

**Comment by quartzmo** [comment]
@blowmage Yes, I think that is a good idea.

**Comment by quartzmo** [comment]
> if content_type isn't specified in PUT then the signature will not match.

@blowmage Does this need to be added to google-cloud-storage docs?

**Comment by blowmage** [comment]
> Does this need to be added to google-cloud-storage docs?

Up to you. I think the current docs are fine, but we have different opinions on the role of docs. :) If you think otherwise I will support any changes you think best.

**Comment by AndrewVos** [comment]
Well you don't really expect http requests when you're newing and object. Maybe that's just me.

Is content_type a required parameter? If it isn't then it probably should be...

**Comment by AndrewVos** [comment]
@blowmage even if the GCS docs do state that clearly it's not where I would be looking if this gem wasn't working for me. I had to search around and found the answer on stackoverflow. If the param is required the people wouldn't have this problem.

Also the problem with doing these requests when you new up the bucket is that in Rails you would have to somehow store a singleton which isn't very nice and is probably unexpected behaviour for most.

**Comment by blowmage** [comment]
> Well you don't really expect http requests when you're newing and object. Maybe that's just me.

Pretty much every object in Google Cloud has an associated API call. This is how we know information about the bucket, such as whether it exists.

> Is content_type a required parameter? If it isn't then it probably should be...

The `content_type` argument is not required because there are URLs that can be generated that do not specify it. For example, a `GET` request usually does not specify a content type.

I did find one code example that was using `PUT` but did not specify a `content_type`, and I have updated the code example to include it.

> Also the problem with doing these requests when you new up the bucket is that in Rails you would have to somehow store a singleton which isn't very nice and is probably unexpected behaviour for most.

You may want to reconsider that. You will likely benefit from memoizing a Google Cloud storage objects on the thread. Every time you call `Google::Cloud::Storage.new` API calls are made to verify your credentials and create an OAuth token. You may also have additional overhead due to establishing a TCP connection, TSL handshake, DNS lookups, etc.

**Comment by AndrewVos** [comment]
> Pretty much every object in Google Cloud has an associated API call. This is how we know information about the bucket, such as whether it exists.

Sure, but what I'm trying to get at here is that it seems to be a convention in Ruby to not do anything much other than setting ivars in `initialize`. Maybe that's just me, but I was 100% sure it was `signed_url` causing the request, because I don't expect calling `new` on something to be doing anything slow.

Regarding `content_type` maybe raising when it is actually required and isn't present depending on the method might be more user friendly? No real opinions on that :)

**Comment by quartzmo** [comment]
Is this issue resolved by #1362?

**Comment by blowmage** [comment]
I don't know if this issue is completely resolved by #1362. It seems that what @AndrewVos is looking for is a way to sign urls completely independently from all the other functionality of google-cloud-storage. Even using googleauth's default credentials might be too much, since that also involves API calls to establish an OAuth token.

I suspect the only way to consider this issue resolved is create a separate signing gem that looks similar to his [`GoogleStorage`](https://github.com/AndrewVos/photo_manager/blob/master/app/lib/google_storage.rb) class.

All that said, I think we can close this issue. Just not sure we can consider it resolved. Thoughts?

**Comment by AndrewVos** [comment]
Yeah that's definitely what I was looking for @blowmage and I just ended up using some code from that link I pasted earlier.

I think this issue is resolved, but perhaps it might be useful to somehow explain in the documentation that the bucket class should be memoized? I can see people copying and pasting the example code in (which instantiates a bucket) and ending up with the belief that Google Cloud Storage requires a request to build a signed url. This almost turned me off from using GCS entirely, which would have been a shame!

Anyway, closing this because it's resolved for me personally. Thanks for your help!
